includes/multi-currency/MultiCurrency.php

```diff
@@ -1184,27 +1154,106 @@ private function register_admin_scripts() {
 		);
 	}
 
+	/**
+	 * Validates the given currency code.
+	 *
+	 * @param   string $currency_code  The currency code to check validity.
+	 *
+	 * @return  string|false  Returns back the currency code in uppercase letters if it's valid, or `false` if not.
+	 */
+	public function validate_currency_code( $currency_code ) {
+		return in_array( strtoupper( $currency_code ), array_keys( $this->available_currencies ), true )
+		? strtoupper( $currency_code )
+		: false;
+	}
+
+	/**
+	 * Get simulation params from querystring and activate when needed
+	 *
+	 * @return  void
+	 */
+	public function possible_simulation_activation() {
+		// This is required in the MC onboarding simulation iframe.
+		$this->simulation_params = $this->get_multi_currency_onboarding_simulation_variables();
+		if ( ! $this->is_simulation_enabled() ) {
+			return;
+		}
+		// Modify the page links to deliver required params in the simulation.
+		$this->add_simulation_params_to_preview_urls();
+		$this->simulate_client_currency();
+	}
+
+	/**
+	 * Enables simulation of client browser currency.
+	 *
+	 * @return  void
+	 */
+	public function simulate_client_currency() {
+		if ( ! $this->simulation_params['enable_auto_currency'] || ! $this->simulation_params['client_currency'] ) {
+			return;
+		}
+		$simulation_currency = $this->simulation_params['client_currency'];
+
+		// Simulate client currency from geolocation.
+		add_filter(
+			'wcpay_multi_currency_get_selected_currency',
+			function( $selected_currency ) use ( $simulation_currency ) {
+				return $simulation_currency;
+			}
+		);
+
+		// Always display the notice on simulation screen.
+		add_action( 'wp_footer', [ $this, 'display_geolocation_currency_update_notice' ] );
```

Comment:
Just a quick question: would it conflict with line 677?
https://github.com/Automattic/woocommerce-payments/pull/2923/files#diff-c722c76bb4254c6fe1a32099d4ea5fa641addb44c5cd685a164e8baaff879e1aR677

Reply:
Don't know, I didn't think about it since you pointed it out here :) You're right, the filter might've been registered twice, then there should be two notice's overlapped. I'll check in the markup if it works that way. Thanks for pointing that out!

Reply:
Fixed.